sphericart/include/sphericart_cuda.hpp

```diff
@@ -0,0 +1,121 @@
+/** \file sphericart_cuda.hpp
+ *  Defines the CUDA API for `sphericart`.
+ */
+#ifndef SPHERICART_CUDA_HPP
+#define SPHERICART_CUDA_HPP
+
+#include <mutex>
+
+#define _SPHERICART_INTERNAL_IMPLEMENTATION
+#include "sphericart.hpp"
+
+namespace sphericart {
+
+namespace cuda {
+
+class CudaSharedMemorySettings {
+  public:
+    CudaSharedMemorySettings()
+        : scalar_size_(0), l_max_(-1), grid_dim_x_(-1), grid_dim_y_(-1),
+          requires_grad_(false), requires_hessian_(false) {}
+
+    /**
+     * Host function to check whether the kernel launch parameters and l_max
+     * value exceeds the default amount of shared memory available on the card.
+     * If true, the function will attempt to increase the shared memory
+     * allocation.
+     *
+     * @param scalar_size
+     *        The size of the scalar type of the input coodinates and output
+     * spherical harmonics (4 or 8 bytes).
+     * @param l_max
+     *        The maximum degree of the spherical harmonics to be calculated.
+     * @param GRID_DIM_X
+     *        The size of the threadblock in the x dimension.
+     * @param GRID_DIM_Y
+     *        The size of the threadblock in the y dimension.
+     * @param gradients
+     *        If we are computing of the first-order derivatives.
+     * @param hessian
+     *        If we are computing of the second-order derivatives.
+     */
+    bool update_if_required(size_t scalar_size, int64_t l_max,
+                            int64_t GRID_DIM_X, int64_t GRID_DIM_Y,
+                            bool gradients, bool hessian);
+
+  private:
+    int64_t l_max_;
+    int64_t grid_dim_x_;
+    int64_t grid_dim_y_;
+    bool requires_grad_;
+    bool requires_hessian_;
+    size_t scalar_size_;
+};
+
+/**
+ * A spherical harmonics calculator.
+ *
+ * It handles initialization of the prefactors upon initialization and it
+ * stores the buffers that are necessary to compute the spherical harmonics
+ * efficiently.
+ */
+template <typename T> class SphericalHarmonics {
+  public:
+    /** Initialize the SphericalHarmonics class setting maximum degree and
+     * normalization
+     *
+     *  @param l_max
+     *      The maximum degree of the spherical harmonics to be calculated.
+     *  @param normalized
+     *      If `false` (default) computes the scaled spherical harmonics, which
+     * are homogeneous polynomials in the Cartesian coordinates of the input
+     * points. If `true`, computes the normalized spherical harmonics that are
+     * evaluated on the unit sphere. In practice, this simply computes the
+     * scaled harmonics at the normalized coordinates \f$(x/r, y/r, z/r)\f$, and
+     * adapts the derivatives accordingly.
+     */
+    SphericalHarmonics(size_t l_max, bool normalized = false);
+
```

Comment:
I think we need to explicitly define the move and copy constructors here. Move constructor should be fine, the copy one should create a new mutex.


```suggestion
    SphericalHarmonics(const SphericalHarmonics& other);
    SphericalHarmonics(SphericalHarmonics&& other);
    
    SphericalHarmonics& operator=(const SphericalHarmonics& other);
    SphericalHarmonics& operator=(SphericalHarmonics&& other);
```

Reply:
we'd also need to add these to the main `sphericart.hpp` class as well then.